cdap-spark-core-base/src/main/java/co/cask/cdap/app/runtime/spark/classloader/SparkRunnerClassLoader.java

```diff
@@ -75,9 +84,70 @@
 
   public SparkRunnerClassLoader(URL[] urls, @Nullable ClassLoader parent, boolean rewriteYarnClient) {
     super(urls, parent);
+    // Copy from URLClassLoader, which also uses WeakHashMap
+    this.closeables = new WeakHashMap<>();
+    this.closeablesLock = new ReentrantLock();
     this.rewriter = new SparkClassRewriter(name -> ClassLoaders.openResource(this, name), rewriteYarnClient);
   }
 
+  @Override
+  public InputStream getResourceAsStream(String name) {
+    URL url = getResource(name);
+    if (url == null) {
+      return null;
+    }
+
+    try {
+      URLConnection urlConn = url.openConnection();
+      urlConn.setUseCaches(false);
+      InputStream is = urlConn.getInputStream();
+      closeablesLock.lock();
+      try {
+        closeables.putIfAbsent(is, null);
+      } finally {
+        closeablesLock.unlock();
+      }
+
+      return is;
+
+    } catch (IOException e) {
+      return null;
+    }
+  }
+
+  @Override
+  public void close() throws IOException {
+    IOException ex = null;
+    try {
+      super.close();
+    } catch (IOException e) {
+      ex = e;
+    }
+
+    // Close all the InputStreams acquired via getResourceAsStream
+    closeablesLock.lock();
+    try {
```

Comment:
so this is to make sure all these streams are closed, even if the caller of getResourceAsStream() does not close it?

Reply:
Yes. It is essentially the same logic as in the Java URLClassLoader class, but it turn off the URL connection caching in the getResourceAsStream method.

Reply:
+1